core/src/main/java/org/apache/calcite/rex/RexUtil.java

```diff
@@ -302,7 +303,16 @@ public static RexNode removeCast(RexNode e) {
     for (;;) {
       switch (e.getKind()) {
       case CAST:
+        RelDataType destType = e.getType();
         e = ((RexCall) e).operands.get(0);
+        if (!SqlTypeUtil.canCastFrom(destType, e.getType(), true)) {
+          throw new UnsupportedOperationException(
```

Comment:
AFAIK, the `removeCast` doesn't say that it would only remove the redundant casts, it just remove `any` casts, so i don't think we should add a `SqlTypeUtil.canCastFrom` check here and throws exception. It is the invoker that should apply these rules.

Reply:
PR updated.
I add the type check for cast in `InputUsageFinder`.

Reply:
@danny0405 , how about this time, check cast in `InputUsageFinder`?